core/src/env/api.rs

```diff
@@ -12,127 +12,439 @@
+    })
+}
+
+/// Invokes a call to the runtime.
+///
+/// # Note
+///
+/// The call is not guaranteed to execute immediately but might be deferred
+/// to the end of the contract execution.
+///
+/// # Errors
 ///
-/// Upon success returns the account ID of the newly created smart contract.
-pub fn create<T>(
+/// - If the called runtime function does not exist.
+pub fn invoke_runtime<T>(params: &T::Call) -> Result<()>
+where
+    T: EnvTypes,
+{
+    <EnvInstance as OnInstance>::on_instance(|instance| {
+        TypedEnv::invoke_runtime::<T>(instance, params)
+    })
+}
+
+/// Invokes a contract message.
+///
+/// # Errors
+///
+/// - If the called contract does not exist.
+/// - If the called contract is a tombstone.
+/// - If arguments passed to the called contract message are invalid.
+/// - If the called contract execution has trapped.
+/// - If the called contract ran out of gas upon execution.
+pub fn invoke_contract<T>(params: &CallParams<T, ()>) -> Result<()>
+where
+    T: EnvTypes,
+{
+    <EnvInstance as OnInstance>::on_instance(|instance| {
+        TypedEnv::invoke_contract::<T>(instance, params)
+    })
+}
+
+/// Evaluates a contract message and returns its result.
+///
+/// # Errors
+///
+/// - If the called contract does not exist.
+/// - If the called contract is a tombstone.
+/// - If arguments passed to the called contract message are invalid.
+/// - If the called contract execution has trapped.
+/// - If the called contract ran out of gas upon execution.
+/// - If the returned value failed to decode properly.
+pub fn eval_contract<T, R>(params: &CallParams<T, ReturnType<R>>) -> Result<R>
+where
+    T: EnvTypes,
+    R: scale::Decode,
+{
+    <EnvInstance as OnInstance>::on_instance(|instance| {
+        TypedEnv::eval_contract::<T, R>(instance, params)
+    })
+}
+
+/// Instantiates another contract.
+///
+/// # Errors
+///
+/// - If the code hash is invalid.
+/// - If the arguments passed to the instantiation process are invalid.
+/// - If the instantiation process traps.
+/// - If the instantiation process runs out of gas.
+/// - If given too few endowment.
+/// - If the returned account ID failed to decode properly.
+pub fn create_contract<T, C>(params: &InstantiateParams<T, C>) -> Result<T::AccountId>
+where
+    T: EnvTypes,
+{
+    <EnvInstance as OnInstance>::on_instance(|instance| {
+        TypedEnv::create_contract::<T, C>(instance, params)
+    })
+}
+
+/// Restores a smart contract in tombstone state.
+///
+/// # Params
+///
+/// - `account_id`: Account ID of the to-be-restored contract.
+/// - `code_hash`: Code hash of the to-be-restored contract.
+/// - `rent_allowance`: Rent allowance of the restored contract
+///                     upon successful restoration.
+/// - `filtered_keys`: Storage keys that will be removed for the tombstone hash
+///                    match calculation that decide whether the original contract
+///                    storage and the storage of the restorer contract are equal.
```

Comment:
See suggestions in another copy of this comment. How can we make sure these comments don't fall out of sync? Perhaps just have the comment in one place and then link to it?

Reply:
I also thought about this problem. I think linking of doc comments is a good idea here. :+1: